Sources/BigInt/BigInt.swift

```diff
@@ -0,0 +1,816 @@
+      lhs.words.removeFirst(Int(fullWords))
+    } else {
+      lhs = isNegative ? -1 : 0
+      return
+    }
+
+    if remainder > 0 {
+      let mask = ~(UInt.max << remainder)
+      var value: UInt = 0
+      for i in stride(from: lhs.words.count - 1, through: 0, by: -1) {
+        let temp = lhs.words[i] & mask
+        lhs.words[i] >>= remainder
+        lhs.words[i] |= (value << (UInt.bitWidth - Int(remainder)))
+        value = temp
+      }
+    }
+
+    if isNegative {
+      lhs.words[lhs.words.count - 1] |= (UInt.max << (UInt.bitWidth - Int(remainder)))
+    }
+
+    BigInt._dropExcessWords(words: &lhs.words)
+  }
+
+  @inlinable
+  public func quotientAndRemainder(dividingBy rhs: BigInt) -> (quotient: BigInt, remainder: BigInt) {
+    return BigInt._div(lhs: self, rhs: rhs)
+  }
+
+  @inlinable
+  public func signum() -> BigInt {
+    if _isNegative {
+      return -1
+    } else if self == 0 {
+      return 0
+    }
+
+    return 1
+  }
+}
+
+// MARK: -
+
+extension BigInt {
+
+  private static func _findQhat(
+    high: UInt,
+    low: UInt.Magnitude,
+    divisor: UInt,
+    nextVdigit: UInt,
+    nextUdigit: UInt
+  ) -> UInt {
+    var (qhat, rhat) = divisor.dividingFullWidth((high, low))
+
+    if high >= divisor { // This means qhat >= b
+      qhat = UInt.max
+    }
+
+    let (tempHigh, tempLow) = qhat.multipliedFullWidth(by: nextVdigit)
+    var (rtempHigh, rtempLow) = rhat.multipliedFullWidth(by: UInt.max)
+    var overflow = false
+    (rtempLow, overflow) = rtempLow.addingReportingOverflow(1)
+    if overflow {
+      rtempHigh += 1
+    }
+
+    (rtempLow, overflow) = rtempLow.addingReportingOverflow(nextUdigit)
+    if overflow {
+      rtempHigh += 1
+    }
+
+    while true {
+      if (tempHigh > rtempHigh) || ((tempHigh == rtempHigh) && (tempLow > rtempLow)) {
+        qhat -= 1
+        (rhat, overflow) = rhat.addingReportingOverflow(divisor)
+        if !overflow {
+          continue
+        } else {
+          break
+        }
+      } else {
+        break
+      }
+    }
+
+    return qhat
+  }
+
+  @usableFromInline
+  internal static func _div(lhs: BigInt, rhs: BigInt) -> (quotient: BigInt, remainder: BigInt) {
+    precondition(rhs != _digits[0], "Division by zero error!")
+
+    if lhs.words.count == 1, rhs.words.count == 1 {
+      let (quot, rem) = Int(bitPattern: lhs.words[0]).quotientAndRemainder(dividingBy: Int(bitPattern: rhs.words[0]))
+      return (BigInt(words: [UInt(bitPattern: quot)]), BigInt(words: [UInt(bitPattern: rem)]))
+    }
+
+    let lhsIsNeg = lhs._isNegative
+    let rhsIsNeg = rhs._isNegative
+
+    var lhsWords = lhsIsNeg ? (-lhs).words : lhs.words
+    var rhsWords = rhsIsNeg ? (-rhs).words : rhs.words
+
+    while rhsWords[rhsWords.endIndex - 1] == 0, rhsWords.count > 2 {
+      rhsWords.removeLast()
+    }
+
+    if rhsWords.count > lhsWords.count { return (0, lhs) }
+
+    if rhsWords.count == 1 {
+      rhsWords.append(0)
+    }
+
+    func normalize(x: UInt) -> UInt {
```

Comment:
Is there a reason this doesn't just use `leadingZeroBitCount`?

Reply:
Because Knuth didn't have access to that for his book, and I was copying fairly literally! I'll fix that.